pkg/controllerhelpers/handlers.go

```diff
@@ -89,29 +89,41 @@ func (h *Handlers[T]) EnqueueAll(depth int, untypedObj kubeinterfaces.ObjectInte
 	}
 }
 
-func (h *Handlers[QT]) EnqueueOwner(depth int, obj kubeinterfaces.ObjectInterface, operation HandlerOperationType) {
-	controllerRef := metav1.GetControllerOf(obj)
-	if controllerRef == nil {
-		return
-	}
+func (h *Handlers[QT]) EnqueueOwnerIf(precondition func(QT) bool) EnqueueFuncType {
+	return func(depth int, obj kubeinterfaces.ObjectInterface, operation HandlerOperationType) {
+		controllerRef := metav1.GetControllerOf(obj)
+		if controllerRef == nil {
+			return
+		}
 
-	if controllerRef.Kind != h.gvk.Kind {
-		return
-	}
+		if controllerRef.Kind != h.gvk.Kind {
+			return
+		}
 
-	owner, err := h.getterLister.Get(obj.GetNamespace(), controllerRef.Name)
-	if err != nil {
-		utilruntime.HandleError(err)
-		return
+		owner, err := h.getterLister.Get(obj.GetNamespace(), controllerRef.Name)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+
+		if owner.GetUID() != controllerRef.UID {
+			utilruntime.HandleError(err)
+			return
+		}
+
+		if precondition(owner) {
```

Comment:
actually similarly to
https://github.com/scylladb/scylla-operator/blob/c5c6c6ffabcacf504e7f39263da377a755f276a1/pkg/controller/nodeconfigpod/controller.go#L202-L216
would it be more generic to customize the enqueue function?

Reply:
It would, and that's what I wanted to do initially, but with `enqueueOwner` you'd have to duplicate the entire logic of extracting the owner.

Reply:
i meant to plumb `enqueue` func, not `enqueueOwner` although that would need to pass it as well

Reply:
I wanted to avoid duplication a version of the function in every variant (enqueue*, enqueueOwner*), see
https://github.com/tnozicka/scylla-operator/commit/803da0be8c47cd7f66fffed1edb2cf5d4c0ca568

Reply:
I see, thanks. Would you get rid of the `WithFilter` func though? I left it in as it allows for adding custom enqueue funcs without duplicating code.

Reply:
The default rule is we don't leave the functions that are not used and bring them when needed.

I don't think it brings simplification because in this case, where the filtering isn't done in the middle, a composition can easily replace it, if not be even more readable:
```
func myEnqueue() {
 // filter
 // call enqueue
}
```
the withFilter could actually be modeled this way as well

Reply:
You're right, it's a bit cleaner to compose it this way. Updated. I hope we're on the same page now and we can merge it?

Reply:
> I don't think it brings simplification because in this case, where the filtering isn't done in the middle, a composition can easily replace it

My point was that the with filter isn't needed at all
```
func (ncpc *Controller) enqueueScyllaPod(depth int, obj kubeinterfaces.ObjectInterface, op controllerhelpers.HandlerOperationType) {
	if !controllerhelpers.IsScyllaPod(pod.(*corev1.Pod)) {
		return
	}

	ncpc.handlers.Enqueue(depth+1, obj, op)
}
```

Reply:
It's not *needed*, but it abstracts away the depth increment nicely